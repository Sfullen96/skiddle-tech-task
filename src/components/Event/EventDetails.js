import React from 'react';
import PropTypes from 'prop-types';
import { withRouter, Link } from 'react-router-dom';
import { Container, Row, Col, Figure, Badge } from 'react-bootstrap';
import moment from 'moment';

import './EventDetails.scss';

const EventDetails = ({
  image,
  name,
  date,
  venue,
  dateStart,
  dateEnd,
  minAge,
  description,
  entryPrice,
  artists,
  genres,
}) => {
  return (
    <Container className="event">
      <Row>
        <Col xs="12" sm="3" className="event__image-container">
          <img
            src={image}
            alt={`${name}`}
            className="figure-img img-fluid img-thumbnail event__image"
          />
        </Col>
        <Col xs="12" sm="9" className="event__main-info">
          <h1 className="event__title">{name}</h1>
          <p className="event__date">
            <i className="fa fa-calendar-alt" />{' '}
            {moment(date, 'YYYY-MM-DD').format('dddd Do MMMM YYYY')}
          </p>
          <p className="event__venue">
            <i className="fa fa-map-marker-alt" /> {venue.name}, {venue.town}
          </p>
          <p className="event__time">
            <i className="fa fa-clock" /> {moment(dateStart, 'YYYY-MM-DD HH:mm:ss').format('H:mma')}{' '}
            - {moment(dateEnd, 'YYYY-MM-DD HH:mm:ss').format('H:mma')}
          </p>
          <p className="event__age-limit">
            <i className="fa fa-user" />{' '}
            {minAge === '0' ? 'No Age Restrictions' : `Minimum Age ${minAge}`}
          </p>
          {genres && genres.length ? (
            <p className="event__tags">
              <i className="fa fa-tag" />{' '}
              {genres.map(genre => (
                <Badge key={genre.genreid} pill variant="info" className="event__tag">
                  {genre.name}
                </Badge>
              ))}
            </p>
          ) : (
            ''
          )}
        </Col>
      </Row>
      <hr />
      <Row>
        <Col>
          <h2 className="event__more-info">More Information</h2>
          <p className="event__description">{description}</p>
        </Col>
      </Row>
      <hr />
      <Row>
        <Col>
          <h3 className="event__price-heading">Entry Prices</h3>
          <p className="event__price">{entryPrice}</p>
        </Col>
      </Row>
      <hr />
      <Row>
        <Col>
          <h2 className="event__more-info">Where is it?</h2>
          <p className="event__address">
            {venue.name}, {venue.town} {venue.postcode}
          </p>
        </Col>
      </Row>
      <hr />
      <Row className="event__artists">
        <Col>
          <h2 className="event__artists-header">Artists</h2>
          <Row>
            {!artists.length ? (
              <Col>
                <p>This is not a musical event</p>
              </Col>
            ) : (
              artists.map(artist => (
                <Col xs="12" sm="4" md="3" key={artist.artistid}>
                  <Figure className="event__artist">
                    <Figure.Image
                      src={artist.image}
                      thumbnail
                      className="event__artist-image"
                      alt={artist.name}
                    />
                    <Link to={`/artist/${artist.artistid}`}>
                      {' '}
                      <Figure.Caption className="event__artist-name">{artist.name}</Figure.Caption>
                    </Link>
                  </Figure>
                </Col>
              ))
            )}
          </Row>
        </Col>
      </Row>
    </Container>
  );
};

EventDetails.propTypes = {
  image: PropTypes.string.isRequired,
  name: PropTypes.string.isRequired,
  date: PropTypes.string.isRequired,
  venue: PropTypes.shape({
    name: PropTypes.string,
    town: PropTypes.string,
    postcode: PropTypes.string,
  }).isRequired,
  dateStart: PropTypes.string.isRequired,
  dateEnd: PropTypes.string.isRequired,
  minAge: PropTypes.string.isRequired,
  description: PropTypes.string.isRequired,
  entryPrice: PropTypes.string.isRequired,
  artists: PropTypes.arrayOf(
    PropTypes.shape({
      artistid: PropTypes.string,
      name: PropTypes.string,
      image: PropTypes.string,
    }),
  ).isRequired,
  genres: PropTypes.arrayOf(
    PropTypes.shape({
      genreid: PropTypes.string,
      name: PropTypes.string,
    }),
  ).isRequired,
};

export default withRouter(EventDetails);
