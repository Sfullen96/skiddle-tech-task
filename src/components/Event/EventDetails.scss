@import '../../globals';

.event {
  background-color: #fff;
  padding-top: 40px;
  padding-bottom: 40px;
  &__image {
    border: 3px solid $primary-border;
    width: 100%;
  }
  &__title {
    font-weight: 800;
    font-size: 28px;
    margin-bottom: 25px;
  }
  &__venue,
  &__date,
  &__time,
  &__age-limit,
  &__tags {
    margin-bottom: 10px;
    font-size: 20px;
  }
  &__tag {
    margin-right: 8px;
    background-color: $primary-background;
    padding: 4px 15px;
  }
  &__description,
  &__price,
  &__address {
    font-size: 18px;
  }
  &__more-info,
  &__artists-header,
  &__price-heading {
    font-size: 22px;
    font-weight: 800;
  }
  &__artist {
    text-align: center;
  }
  &__artist-image {
    width: 100%;
    border: 3px solid $primary-border;
  }
  &__artist-name {
    font-weight: 800;
    font-size: 18px;
  }
}

// Mobile
@media (min-width: 320px) and (max-width: 480px) {
  .event {
    &__title {
      font-size: 24px;
      margin-bottom: 10px;
    }
    &__venue,
    &__date,
    &__time,
    &__age-limit,
    &__tags {
      font-size: 16px;
    }
    &__description,
    &__price,
    &__address {
      font-size: 16px;
    }
  }
}
