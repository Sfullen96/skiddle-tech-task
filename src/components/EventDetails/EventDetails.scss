@import '../../globals';

.event {
  background-color: #fff;
  padding-top: 40px;
  padding-bottom: 40px;
  &__image {
    border: 3px solid $primary-border;
    width: 100%;
  }
  &__title {
    font-weight: 800;
    font-size: 26px;
    margin-bottom: 25px;
  }
  &__venue,
  &__date,
  &__time,
  &__age-limit {
    margin-bottom: 10px;
    font-size: 20px;
  }
  &__description,
  &__price {
    font-size: 18px;
  }
  &__more-info,
  &__artists-header,
  &__price-heading {
    font-size: 22px;
    font-weight: 800;
  }
}
