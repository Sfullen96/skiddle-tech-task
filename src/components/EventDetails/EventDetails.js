import React, { useEffect, useState } from 'react';
import PropTypes from 'prop-types';
import { withRouter } from 'react-router-dom';
import { Container, Row, Col, Figure } from 'react-bootstrap';
import moment from 'moment';

import './EventDetails.scss';

import { Loader } from '..';
import { getEvent as getEventService } from '../../services';

const EventDetails = ({
  match: {
    params: { id },
  },
}) => {
  const [event, setEvent] = useState(null);

  useEffect(() => {
    const getEvent = async _id => {
      const data = await getEventService(_id);
      setEvent(data);
    };
    getEvent(id);
  }, [id]);

  if (!event) {
    return <Loader />;
  }

  const {
    largeimageurl,
    eventname,
    date,
    venue,
    dateStart,
    dateEnd,
    MinAge,
    description,
    entryprice,
    artists,
  } = event;

  return (
    <Container className="event">
      <Row>
        <Col xs="12" sm="3" className="event__image-container">
          <img
            src={largeimageurl}
            alt={`${eventname}`}
            className="figure-img img-fluid img-thumbnail event__image"
          />
        </Col>
        <Col xs="12" sm="9" className="event__main-info">
          <h1 className="event__title">{eventname}</h1>
          <p className="event__date">
            <i className="fa fa-calendar-alt" /> {moment(date).format('dddd Do MMMM YYYY')}
          </p>
          <p className="event__venue">
            <i className="fa fa-map-marker-alt" /> {venue.name}, {venue.town}
          </p>
          <p className="event__time">
            <i className="fa fa-clock" /> {moment(dateStart).format('H:mma')} -{' '}
            {moment(dateEnd).format('H:mma')}
          </p>
          <p className="event__age-limit">
            <i className="fa fa-user" />{' '}
            {MinAge === '0' ? 'No Age Restrictions' : `Minimum Age ${MinAge}`}
          </p>
        </Col>
      </Row>
      <hr />
      <Row>
        <Col>
          <h2 className="event__more-info">More Information</h2>
          <p className="event__description">{description}</p>
          <h3 className="event__price-heading">Entry Prices</h3>
          <p className="event__price">{entryprice}</p>
        </Col>
      </Row>
      <hr />
      <Row>
        <Col>
          <h2 className="event__artists-header">Artists</h2>
          {!artists.length
            ? 'This is not a musical event'
            : artists.map(artist => (
                <Figure>
                  <Figure.Image src={artist.image} thumbnail />
                  <Figure.Caption>{artist.name}</Figure.Caption>
                </Figure>
              ))}
        </Col>
      </Row>
    </Container>
  );
};

EventDetails.propTypes = {
  match: PropTypes.shape({
    params: PropTypes.shape({
      id: PropTypes.number.isRequired,
    }).isRequired,
  }).isRequired,
};

export default withRouter(EventDetails);
