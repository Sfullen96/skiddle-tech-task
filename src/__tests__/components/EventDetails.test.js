import React from 'react';
import { shallow } from 'enzyme';

import EventDetails from '../../components/Event/EventDetails';

describe('<EventDetails />', () => {
  let wrapper;
  beforeEach(() => {
    wrapper = shallow(
      <EventDetails.WrappedComponent
        image="testimage"
        name="Courteeners"
        date="22-06-19"
        venue={{ name: 'Heaton Park', town: 'Manchester', postcode: 'M253EX' }}
        dateStart="22-06-19 10:00:00"
        dateEnd="22-06-19 23:00:00"
        minAge="0"
        description="Description of this event"
        entryPrice="tbc"
        artists={[
          {
            artistid: '1',
            name: 'Courteeners',
            image: 'courteeners.png',
          },
          {
            artistId: '2',
            name: 'James',
            image: 'james.png',
          },
        ]}
        genres={[
          {
            genreid: '1',
            name: 'Indie',
          },
          {
            genreid: '2',
            name: 'Rock',
          },
        ]}
      />,
    );
  });

  it('should match the snaphot', () => expect(wrapper).toMatchSnapshot());

  it('should render without crashing', () => {
    expect(wrapper.find('Container.event').length).toEqual(1);
  });

  it('should render the event name', () => {
    wrapper.setProps({ name: 'Test Name' });
    expect(wrapper.find('.event__title').text()).toEqual('Test Name');
  });

  it('should show the image of the event, and it should have an appropriate alt', () => {
    wrapper.setProps({ image: 'testimage.png', name: 'Test Event' });
    expect(wrapper.find('.event__image').props().src).toEqual('testimage.png');
    expect(wrapper.find('.event__image').props().alt).toEqual('Test Event');
  });

  it('should render the event information, including date, venue, times and age restrictions', () => {
    wrapper.setProps({
      date: '2019-06-15',
      dateStart: '2019-06-15 10:00:00',
      dateEnd: '2019-06-15 23:00:00',
      venue: { name: 'Manchester Arena', town: 'Manchester', postcode: 'M46WG' },
      minAge: '0',
    });
    expect(wrapper.find('.event__date').text()).toEqual(' Saturday 15th June 2019');
    expect(wrapper.find('.event__venue').text()).toEqual(' Manchester Arena, Manchester');
    expect(wrapper.find('.event__time').text()).toEqual(' 10:00am - 23:00pm');
    expect(wrapper.find('.event__age-limit').text()).toEqual(' No Age Restrictions');

    wrapper.setProps({ minAge: '18' });
    expect(wrapper.find('.event__age-limit').text()).toEqual(' Minimum Age 18');
  });

  it('should render the correct tags', () => {
    wrapper.setProps({ genres: [{ genreid: '1', name: 'Indie' }, { genreid: '2', name: 'Rock' }] });
    expect(wrapper.find('.event__tags').find('Badge').length).toEqual(2);
    expect(
      wrapper
        .find('.event__tags')
        .find('Badge')
        .first()
        .text(),
    ).toEqual('Indie');
    expect(
      wrapper
        .find('.event__tags')
        .find('Badge')
        .at(1)
        .text(),
    ).toEqual('Rock');

    wrapper.setProps({ genres: [] });
    expect(wrapper.find('.event__tags').length).toEqual(0);
  });

  it('should render the description', () => {
    expect(wrapper.find('.event__description').text()).toEqual('Description of this event');
  });

  it('should render the entry prices', () => {
    expect(wrapper.find('.event__price').text()).toEqual('tbc');
  });

  it('should render the event address', () => {
    expect(wrapper.find('.event__address').text()).toEqual('Heaton Park, Manchester M253EX');
  });

  it('should not try and render artists if there are none', () => {
    wrapper.setProps({ artists: [] });
    expect(wrapper.find('.event__artist').length).toEqual(0);
    expect(wrapper.find('.event__artists p').text()).toEqual('This is not a musical event');
  });

  it('should render artists', () => {
    expect(
      wrapper
        .find('.event__artist')
        .find('.event__artist-name')
        .first()
        .text(),
    ).toEqual('Courteeners');
    expect(
      wrapper
        .find('.event__artist')
        .find('.event__artist-image')
        .first()
        .props().src,
    ).toEqual('courteeners.png');
    expect(
      wrapper
        .find('.event__artist')
        .find('.event__artist-image')
        .first()
        .props().alt,
    ).toEqual('Courteeners');

    expect(
      wrapper
        .find('.event__artist')
        .find('.event__artist-name')
        .at(1)
        .text(),
    ).toEqual('James');
    expect(
      wrapper
        .find('.event__artist')
        .find('.event__artist-image')
        .at(1)
        .props().src,
    ).toEqual('james.png');
    expect(
      wrapper
        .find('.event__artist')
        .find('.event__artist-image')
        .at(1)
        .props().alt,
    ).toEqual('James');
  });
});
