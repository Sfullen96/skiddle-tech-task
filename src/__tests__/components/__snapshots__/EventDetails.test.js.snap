// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<EventDetails /> should match the snaphot 1`] = `ShallowWrapper {}`;
